dht.go

```diff
@@ -275,7 +276,49 @@ func (dht *IpfsDHT) putLocal(key string, rec *recpb.Record) error {
 // on the given peer.
 func (dht *IpfsDHT) Update(ctx context.Context, p peer.ID) {
 	logger.Event(ctx, "updatePeer", p)
-	dht.routingTable.Update(p)
+	if dht.shouldAddPeerToRoutingTable(p) {
+		dht.routingTable.Update(p)
+	}
+}
+
+func (dht *IpfsDHT) shouldAddPeerToRoutingTable(p peer.ID) bool {
+	return dht.hasOutboundConnToPeer(p) ||
+		dht.hasSensibleAddressesForPeer(p)
+}
+
+func (dht *IpfsDHT) hasOutboundConnToPeer(p peer.ID) bool {
+	cons := dht.host.Network().ConnsToPeer(p)
+	for _, c := range cons {
+		if c.Stat().Direction == network.DirOutbound {
+			return true
+		}
+	}
+	return false
+}
+
+func (dht *IpfsDHT) hasSensibleAddressesForPeer(p peer.ID) bool {
+	if dht.peerIsOnSameSubnet(p) {
+		// TODO: for now, we can't easily tell if the peer on our subnet
+		// is dialable or not, so don't discriminate.
+		return true
+	}
+
+	for _, a := range dht.host.Peerstore().Addrs(p) {
```

Comment:
Did we already deal with advertising tons of relay addresses? If not this can get quite expensive.

Reply:
oooh, I think we're mostly past that, but should we check the number of addresses here? If we have a ton, we probably don't want that peer in the routing table...